pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -17,20 +17,59 @@ limitations under the License.
 package persistentvolume
 
 import (
+	"fmt"
 	"time"
 
 	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/client/cache"
 	clientset "k8s.io/kubernetes/pkg/client/clientset_generated/internalclientset"
 	"k8s.io/kubernetes/pkg/cloudprovider"
 	"k8s.io/kubernetes/pkg/controller/framework"
+	"k8s.io/kubernetes/pkg/conversion"
 	"k8s.io/kubernetes/pkg/runtime"
 	vol "k8s.io/kubernetes/pkg/volume"
 	"k8s.io/kubernetes/pkg/watch"
 
 	"github.com/golang/glog"
 )
 
+// Design:
+//
+// The fundamental key to this design is the bi-directional "pointer" between
+// PersistentVolumes (PVs) and PersistentVolumeClaims (PVCs), which is
+// represented here as pvc.Spec.VolumeName and pv.Spec.ClaimRef. The bi-directionality
+// is complicated to manage in a transactionless system, but without it we
+// can't ensure sane behavior in the face of different forms of trouble.  For
+// example, a rogue HA controller instance could end up racing and making
+// multiple bindings that are indistinguishable, resulting in potential data
+// loss.
+//
+// This controller supports pre-bound (by the creator) objects in both
+// directions: a PVC that wants a specific PV or a PV that is reserved for a
+// specific PVC.
+//
+// The binding is two-step process. PV.Spec.ClaimRef is modified first and
+// PVC.Spec.VolumeName second. At any point of this transaction, the PV or PVC
+// can be modified by user or other controller or completelly deleted. Also, two
+// (or more) controllers may try to bind different volumes to different claims
+// at the same time. The controller must recover from any conflicts that may
+// arise from these conditions.
+
+// annWasEverBound annotation applies to PVCs.  It indicates that the lifecycle
```

Comment:
e.g.

```
diff --git a/controller.go b/controller.go
index 311e6f7..cc543f0 100644
--- a/controller.go
+++ b/controller.go
@@ -20,15 +20,15 @@
 // document where things could break down in an active/active situation, but
 // active/active is out of scope for now.
 
-// This annotation applies to PVCs.  It indicates that the lifecycle of the PVC
-// has passed through the initial setup.  This information changes how we
-// interpret some observations of the state of the objects.
-const annWasEverBound = "pv.kubernetes.io/bound-completed"
-
 // This annotation applies to PVs and PVCs.  It indicates that the binding
-// (PV->PVC or PVC->PV) was installed by the controller.  The absence of this
-// annotation means the binding was done by the user (i.e. pre-bound).
-const annBoundByController = "pv.kubernetes.io/bound-by-controller"
+// (PV->PVC or PVC->PV) was completed and by whom.  Any value other than
+// "Controller" means that this controller will not override the binding.  The
+// fact that the binding was completed indicates that the lifecycle of the PVC
+// has passed through the initial setup.  This changes how we interpret some
+// observations of the state of the objects.
+const annBoundBy = "pv.kubernetes.io/bound-by"
+const boundByUnknown = "unknown"
+const boundByController = "pv.kubernetes.io/controller"
 
 // This annotation represents a new field which instructs dynamic provisioning
 // to choose a particular storage class (aka profile).
@@ -43,7 +43,7 @@ const annDynamicallyProvisioned = "pv.kubernetes.io/provisioned-by"
 // This must be async-safe, idempotent, and crash/restart safe, since it
 // happens in a loop as well as on-demand.
 func SyncPVC(pvc *PVClaim) {
-	if !hasAnnotation(pvc, annWasEverBound) {
+	if !hasAnnotation(pvc, annBoundBy) {
 		// This is a new PVC that has not completed binding
 		// OBSERVATION: pvc is "Pending"
 		if pvc.Spec.VolumePtr == nil {
@@ -85,7 +85,7 @@ func SyncPVC(pvc *PVClaim) {
 				if pv.Spec.ClaimPtr == nil {
 					pv.Spec.ClaimPtr = pvc
 					pv.Spec.ClaimPtr.UID = pvc.UID
-					setAnnotation(pv, annBoundByController)
+					setAnnotation(pv, annBoundBy, boundByController)
 				}
 				if err := CommitPV(pv); err != nil {
 					// Nothing was saved; we will fall back into the same
@@ -99,8 +99,7 @@ func SyncPVC(pvc *PVClaim) {
 				}
 				// OBSERVATION: pvc is "Pending", pv is "Bound"
 				pvc.Spec.VolumePtr = pv
-				setAnnotation(pvc, annWasEverBound)
-				setAnnotation(pvc, annBoundByController)
+				setAnnotation(pvc, annBoundBy, boundByController)
 				if err := CommitPVC(pvc); err != nil {
 					// Commit failed; we will handle this partially committed
 					// state in the next call to syncPVC
@@ -126,7 +125,7 @@ func SyncPVC(pvc *PVClaim) {
 				// OBSERVATION: pvc is "Pending", pv is "Available"
 				pv.Spec.ClaimPtr = pvc
 				pv.Spec.ClaimPtr.UID = pvc.UID
-				setAnnotation(pv, annBoundByController)
+				setAnnotation(pv, annBoundBy, boundByController)
 				if err := CommitPV(pv); err != nil {
 					// Retry later.
 					return
@@ -137,7 +136,7 @@ func SyncPVC(pvc *PVClaim) {
 					return
 				}
 				// OBSERVATION: pvc is "Pending", pv is "Bound"
-				setAnnotation(pvc, annWasEverBound)
+				setAnnotation(pvc, annBoundBy, boundByUnknown)
 				if err := CommitPVC(pvc); err != nil {
 					// Retry later.
 					return
@@ -161,7 +160,7 @@ func SyncPVC(pvc *PVClaim) {
 					// Status was not saved. syncPV will set the status
 					return
 				}
-				setAnnotation(pvc, annWasEverBound)
+				setAnnotation(pvc, annBoundBy, boundByUnknown)
 				if err := CommitPVC(pvc); err != nil {
 					// Retry later.
 					return
@@ -175,7 +174,7 @@ func SyncPVC(pvc *PVClaim) {
 			} else {
 				// User asked for a PV that is claimed by someone else
 				// OBSERVATION: pvc is "Pending", pv is "Bound"
-				if !hasAnnotation(pvc, annBoundByController) {
+				if !hasAnnotationValue(pvc, annBoundBy, boundByController) {
 					// User asked for a specific PV, retry later
 					return
 				} else {
@@ -185,7 +184,7 @@ func SyncPVC(pvc *PVClaim) {
 				}
 			}
 		}
-	} else /* hasAnnotation(pvc, annWasEverBound) */ {
+	} else /* hasAnnotation(pvc, annBoundBy) */ {
 		// This PVC has previously been bound
 		// OBSERVATION: pvc is not "Pending"
 		if pvc.Spec.VolumePtr == nil {
@@ -350,7 +349,7 @@ func syncPV(pv *PV) {
 		} else if pvc.Spec.VolumePtr == nil {
 			// This block collapses into a NOP; we're leaving this here for
 			// completeness.
-			if hasAnnotation(pv, annBoundByController) {
+			if hasAnnotationValue(pv, annBoundBy, boundByController) {
 				// The binding is not completed; let PVC sync handle it
 			} else {
 				// Dangling PV; try to re-establish the link in the PVC sync
@@ -392,7 +391,7 @@ func syncPV(pv *PV) {
 				}
 			} else {
 				// This volume is not dynamically provisioned
-				if hasAnnotation(pv, annBoundByController) {
+				if hasAnnotationValue(pv, annBoundBy, boundByController) {
 					// This is part of the normal operation of the controller;
 					// the controller tried to use this volume for a claim but the claim
 					// was fulfilled by another volume.
```